README.md
# Bitset
Bitset is a Go library which represents a **fixed** size sequence of N bits. 
This library is useful to implement a bloom filter on top of, or any place you 
need a bunch of bits to manipulate.

# Installation

> go get github.com/umahmood/bitset

> cd $GOPATH/src/github.com/umahmood/bitset
    
> go test ./...

# Documentation

> http://godoc.org/github.com/umahmood/bitset

# License

See the [LICENSE](LICENSE.md) file for license rights and limitations (MIT).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 2015-10-27
### Changed
- Ran golint on project resulting in minor code changes.

### Added
- Library is now versioned bitset.Version().


bitset_test.go
package bitset

import "testing"

func TestNewReturnsValidBitset(t *testing.T) {
	b := New(100)
	if b == nil {
		t.Errorf("new: error in creating bitset instance.")
	}
}

func TestWordIndex(t *testing.T) {
	in := []uint64{0, 64, 128, 192, 256, 320, 384, 448, 512, 576}
	want := []uint64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
	b := New(100)
	n := 0
	for n < len(in) {
		got := b.wordIndex(in[n])
		if got != want[n] {
			t.Errorf("wordIndex: incorrect index got %d want %d", got, want[n])
		}
		n++
	}
}

func TestPosIndex(t *testing.T) {
	check := func(got, want uint64) {
		if got != want {
			t.Errorf("posIndex: incorrect index got %d want %d", got, want)
		}
	}

	b := New(100)

	check(b.posIndex(0, 0), 0)
	check(b.posIndex(42, 0), 42)
	check(b.posIndex(63, 0), 63)

	check(b.posIndex(64, 1), 0)
	check(b.posIndex(110, 1), 46)
	check(b.posIndex(127, 1), 63)

	check(b.posIndex(128, 2), 0)
	check(b.posIndex(145, 2), 17)
	check(b.posIndex(191, 2), 63)

	check(b.posIndex(192, 3), 0)
	check(b.posIndex(242, 3), 50)
	check(b.posIndex(255, 3), 63)

	check(b.posIndex(256, 4), 0)
	check(b.posIndex(308, 4), 52)
	check(b.posIndex(319, 4), 63)
}

func TestCheckBoundsThrowsPanic(t *testing.T) {
	want := "index out of bounds"
	b := New(100)
	defer func() {
		if got := recover(); got != nil {
			switch got.(type) {
			case string:
				if got != want {
					t.Errorf("checkBounds: got \"%s\" want \"%s\"", got, want)
				}
			}
		}
	}()

	b.checkBounds(200)
}

func TestCheckBoundsDoesNotThrowPanic(t *testing.T) {
	var want error
	b := New(100)
	defer func() {
		if got := recover(); got != want {
			t.Errorf("checkBounds: got %T want %T", got, want)
		}
	}()

	b.checkBounds(0)
	b.checkBounds(63)
	b.checkBounds(127)
}

func TestSizeIsRoundedUpToMultipleOf64(t *testing.T) {
	var in uint64 = 100
	var want uint64 = 128

	b := New(in)
	got := b.Size()

	if got != want {
		t.Errorf("size: mis-match bitset(%d) got %d want %d", in, got, want)
	}
}

func TestSet(t *testing.T) {
	in := []uint64{0, 5, 63, 64, 99, 127}
	want := true
	b := New(100)
	for _, v := range in {
		b.Set(v)
		got := b.Test(v)
		if got != want {
			t.Errorf("set: bit not set b.set(%d) got %t want %t", v, got, want)
		}
	}
}

func TestReset(t *testing.T) {
	in := []uint64{0, 5, 63, 64, 99, 127}
	want := false
	b := New(100)
	for _, v := range in {
		b.Set(v)
		b.Reset(v)
		got := b.Test(v)
		if got != want {
			t.Errorf("reset: bit not set b.set(%d) got %t want %t", v, got, want)
		}
	}
}

func TestFlip(t *testing.T) {
	in := []uint64{0, 63, 64, 127}
	want := true
	b := New(100)
	for _, v := range in {
		b.Flip(v)
		got := b.Test(v)
		if got != want {
			t.Errorf("flip: bit not flipped b.flip(%d) got %t want %t", v, got, want)
		}
	}
}

func TestFlipAndThenUnFlip(t *testing.T) {
	in := []uint64{0, 63, 64, 127}
	want := false
	b := New(100)
	for _, v := range in {
		b.Flip(v) // flip
		b.Flip(v) // un-flip
		got := b.Test(v)
		if got != want {
			t.Errorf("flip: bit not unflipped b.flip(%d) got %t want %t", v, got, want)
		}
	}
}

func TestMethodTestForBitsSetToTrue(t *testing.T) {
	in := []uint64{0, 5, 63, 64, 99, 127}
	want := true
	b := New(100)
	for _, v := range in {
		b.Set(v)
		got := b.Test(v)
		if got != want {
			t.Errorf("test: bit not set b.test(%d) got %t want %t", v, got, want)
		}
	}
}

func TestMethodTestForBitsSetToFalse(t *testing.T) {
	in := []uint64{0, 5, 63, 64, 99, 127}
	want := false
	b := New(100)
	for _, v := range in {
		got := b.Test(v)
		if got != want {
			t.Errorf("test: bit not set b.test(%d) got %t want %t", v, got, want)
		}
	}
}

func TestAllSomeBitsSet(t *testing.T) {
	in := []uint64{0, 63, 64, 127}
	want := false
	b := New(100)
	for _, v := range in {
		b.Set(v)
	}
	got := b.All()
	if got != want {
		t.Errorf("test: all bits set b.All() got %t want %t", got, want)
	}
}

func TestAllNoBitsSet(t *testing.T) {
	want := false
	b := New(100)
	got := b.All()
	if got != want {
		t.Errorf("test: all bits set b.All() got %t want %t", got, want)
	}
}

func TestAllBitsSet(t *testing.T) {
	want := true
	b := New(100)
	var i uint64
	for ; i < b.Size(); i++ {
		b.Set(i)
	}
	got := b.All()
	if got != want {
		t.Errorf("test: all bits not set b.All() got %t want %t", got, want)
	}
}

func TestAnyWithBitsSet(t *testing.T) {
	want := true
	b := New(100)
	b.Set(0)
	b.Set(127)
	got := b.Any()
	if got != want {
		t.Errorf("test: no bits set b.Any() got %t want %t", got, want)
	}
}

func TestAnyNoBitsSet(t *testing.T) {
	want := false
	b := New(100)
	got := b.Any()
	if got != want {
		t.Errorf("test: bits set b.Any() got %t want %t", got, want)
	}
}

func TestNoneNoBitsSet(t *testing.T) {
	want := true
	b := New(100)
	got := b.None()
	if got != want {
		t.Errorf("test: bits set b.None() got %t want %t", got, want)
	}
}

func TestNoneSomeBitsSet(t *testing.T) {
	want := false
	b := New(100)
	b.Set(114)
	b.Set(127)
	got := b.None()
	if got != want {
		t.Errorf("test: no bits set b.None() got %t want %t", got, want)
	}
}

func TestTrueCount(t *testing.T) {
	in := []uint64{0, 10, 63, 64, 110, 127}
	var want uint64 = 5
	b := New(100)
	for _, v := range in {
		b.Set(v)
	}
	b.Reset(10)
	b.Reset(110)
	b.Reset(42) // 42nd bit not set.
	b.Flip(65)
	got := b.TrueCount()
	if got != want {
		t.Errorf("test: count mis-match b.TrueCount() got %d want %d", got, want)
	}
}

func TestCopySrcAndDstBitsAreExact(t *testing.T) {
	a := New(100)
	a.Set(0)
	a.Set(63)
	a.Set(64)
	a.Set(127)
	b := New(200)
	b.Copy(a)
	var i uint64
	for i = 0; i < a.Size(); i++ {
		want := a.Test(i)
		got := b.Test(i)
		if got != want {
			t.Errorf("copy: did not make an exact copy of source bitset, "+
				"position %d got %t want %t", i, got, want)
		}
	}
}

func TestCopySrcAndDstTrueCountAreExact(t *testing.T) {
	a := New(100)
	a.Set(0)
	a.Set(63)
	a.Set(64)
	a.Set(127)
	b := New(200)
	b.Copy(a)
	want := a.TrueCount()
	got := b.TrueCount()
	if got != want {
		t.Errorf("copy: true count mis-match got %d want %d", got, want)
	}
}

func TestCopySrcAndDstSizesAreExact(t *testing.T) {
	a := New(200)
	b := New(100)
	b.Copy(a)
	want := a.Size()
	got := b.Size()
	if got != want {
		t.Errorf("copy: size mis-match got %d want %d", got, want)
	}
}

func TestStringer(t *testing.T) {
	in := []uint64{0, 63, 64, 127}
	want := "1000000000000000000000000000000000000000000000000000000000000001" +
		"1000000000000000000000000000000000000000000000000000000000000001"
	b := New(100)
	for _, v := range in {
		b.Set(v)
	}
	got := b.String()
	if got != want {
		t.Errorf("string: bit pattern mis-match got %s want %s", got, want)
	}
}


bitset_internal_test.go
package bitset

import "testing"

func TestWordIndex(t *testing.T) {
	in := []uint64{0, 64, 128, 192, 256, 320, 384, 448, 512, 576}
	want := []uint64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
	b := New(100)
	n := 0
	for n < len(in) {
		got := b.wordIndex(in[n])
		if got != want[n] {
			t.Errorf("wordIndex: incorrect index got %d want %d", got, want[n])
		}
		n++
	}
}

func TestPosIndex(t *testing.T) {
	check := func(got, want uint64) {
		if got != want {
			t.Errorf("posIndex: incorrect index got %d want %d", got, want)
		}
	}

	b := New(100)

	check(b.posIndex(0, 0), 0)
	check(b.posIndex(42, 0), 42)
	check(b.posIndex(63, 0), 63)

	check(b.posIndex(64, 1), 0)
	check(b.posIndex(110, 1), 46)
	check(b.posIndex(127, 1), 63)

	check(b.posIndex(128, 2), 0)
	check(b.posIndex(145, 2), 17)
	check(b.posIndex(191, 2), 63)

	check(b.posIndex(192, 3), 0)
	check(b.posIndex(242, 3), 50)
	check(b.posIndex(255, 3), 63)

	check(b.posIndex(256, 4), 0)
	check(b.posIndex(308, 4), 52)
	check(b.posIndex(319, 4), 63)
}

func TestCheckBoundsThrowsPanic(t *testing.T) {
	want := "index out of bounds"
	b := New(100)
	defer func() {
		if got := recover(); got != nil {
			switch got.(type) {
			case string:
				if got != want {
					t.Errorf("checkBounds: got \"%s\" want \"%s\"", got, want)
				}
			}
		}
	}()

	b.checkBounds(200)
}

func TestCheckBoundsDoesNotThrowPanic(t *testing.T) {
	var want error
	b := New(100)
	defer func() {
		if got := recover(); got != want {
			t.Errorf("checkBounds: got %T want %T", got, want)
		}
	}()

	b.checkBounds(0)
	b.checkBounds(63)
	b.checkBounds(127)
}

func TestSizeIsRoundedUpToMultipleOf64(t *testing.T) {
	var in uint64 = 100
	var want uint64 = 128

	b := New(in)
	got := b.Size()

	if got != want {
		t.Errorf("size: mis-match bitset(%d) got %d want %d", in, got, want)
	}
}


bitset_test/bitset_test.go
package bitset_test

import (
	"testing"

	"github.com/umahmood/bitset"
)

func TestNewReturnsValidBitset(t *testing.T) {
	b := bitset.New(100)
	if b == nil {
		t.Errorf("new: error in creating bitset instance.")
	}
}

func TestSet(t *testing.T) {
	in := []uint64{0, 5, 63, 64, 99, 127}
	want := true
	b := bitset.New(100)
	for _, v := range in {
		b.Set(v)
		got := b.Test(v)
		if got != want {
			t.Errorf("set: bit not set b.set(%d) got %t want %t", v, got, want)
		}
	}
}

func TestReset(t *testing.T) {
	in := []uint64{0, 5, 63, 64, 99, 127}
	want := false
	b := bitset.New(100)
	for _, v := range in {
		b.Set(v)
		b.Reset(v)
		got := b.Test(v)
		if got != want {
			t.Errorf("reset: bit not set b.set(%d) got %t want %t", v, got, want)
		}
	}
}

func TestFlip(t *testing.T) {
	in := []uint64{0, 63, 64, 127}
	want := true
	b := bitset.New(100)
	for _, v := range in {
		b.Flip(v)
		got := b.Test(v)
		if got != want {
			t.Errorf("flip: bit not flipped b.flip(%d) got %t want %t", v, got, want)
		}
	}
}

func TestFlipAndThenUnFlip(t *testing.T) {
	in := []uint64{0, 63, 64, 127}
	want := false
	b := bitset.New(100)
	for _, v := range in {
		b.Flip(v) // flip
		b.Flip(v) // un-flip
		got := b.Test(v)
		if got != want {
			t.Errorf("flip: bit not unflipped b.flip(%d) got %t want %t", v, got, want)
		}
	}
}

func TestMethodTestForBitsSetToTrue(t *testing.T) {
	in := []uint64{0, 5, 63, 64, 99, 127}
	want := true
	b := bitset.New(100)
	for _, v := range in {
		b.Set(v)
		got := b.Test(v)
		if got != want {
			t.Errorf("test: bit not set b.test(%d) got %t want %t", v, got, want)
		}
	}
}

func TestMethodTestForBitsSetToFalse(t *testing.T) {
	in := []uint64{0, 5, 63, 64, 99, 127}
	want := false
	b := bitset.New(100)
	for _, v := range in {
		got := b.Test(v)
		if got != want {
			t.Errorf("test: bit not set b.test(%d) got %t want %t", v, got, want)
		}
	}
}

func TestAllSomeBitsSet(t *testing.T) {
	in := []uint64{0, 63, 64, 127}
	want := false
	b := bitset.New(100)
	for _, v := range in {
		b.Set(v)
	}
	got := b.All()
	if got != want {
		t.Errorf("test: all bits set b.All() got %t want %t", got, want)
	}
}

func TestAllNoBitsSet(t *testing.T) {
	want := false
	b := bitset.New(100)
	got := b.All()
	if got != want {
		t.Errorf("test: all bits set b.All() got %t want %t", got, want)
	}
}

func TestAllBitsSet(t *testing.T) {
	want := true
	b := bitset.New(100)
	var i uint64
	for ; i < b.Size(); i++ {
		b.Set(i)
	}
	got := b.All()
	if got != want {
		t.Errorf("test: all bits not set b.All() got %t want %t", got, want)
	}
}

func TestAnyWithBitsSet(t *testing.T) {
	want := true
	b := bitset.New(100)
	b.Set(0)
	b.Set(127)
	got := b.Any()
	if got != want {
		t.Errorf("test: no bits set b.Any() got %t want %t", got, want)
	}
}

func TestAnyNoBitsSet(t *testing.T) {
	want := false
	b := bitset.New(100)
	got := b.Any()
	if got != want {
		t.Errorf("test: bits set b.Any() got %t want %t", got, want)
	}
}

func TestNoneNoBitsSet(t *testing.T) {
	want := true
	b := bitset.New(100)
	got := b.None()
	if got != want {
		t.Errorf("test: bits set b.None() got %t want %t", got, want)
	}
}

func TestNoneSomeBitsSet(t *testing.T) {
	want := false
	b := bitset.New(100)
	b.Set(114)
	b.Set(127)
	got := b.None()
	if got != want {
		t.Errorf("test: no bits set b.None() got %t want %t", got, want)
	}
}

func TestTrueCount(t *testing.T) {
	in := []uint64{0, 10, 63, 64, 110, 127}
	var want uint64 = 5
	b := bitset.New(100)
	for _, v := range in {
		b.Set(v)
	}
	b.Reset(10)
	b.Reset(110)
	b.Reset(42) // 42nd bit not set.
	b.Flip(65)
	got := b.TrueCount()
	if got != want {
		t.Errorf("test: count mis-match b.TrueCount() got %d want %d", got, want)
	}
}

func TestCopySrcAndDstBitsAreExact(t *testing.T) {
	a := bitset.New(100)
	a.Set(0)
	a.Set(63)
	a.Set(64)
	a.Set(127)
	b := bitset.New(200)
	b.Copy(a)
	var i uint64
	for i = 0; i < a.Size(); i++ {
		want := a.Test(i)
		got := b.Test(i)
		if got != want {
			t.Errorf("copy: did not make an exact copy of source bitset, "+
				"position %d got %t want %t", i, got, want)
		}
	}
}

func TestCopySrcAndDstTrueCountAreExact(t *testing.T) {
	a := bitset.New(100)
	a.Set(0)
	a.Set(63)
	a.Set(64)
	a.Set(127)
	b := bitset.New(200)
	b.Copy(a)
	want := a.TrueCount()
	got := b.TrueCount()
	if got != want {
		t.Errorf("copy: true count mis-match got %d want %d", got, want)
	}
}

func TestCopySrcAndDstSizesAreExact(t *testing.T) {
	a := bitset.New(200)
	b := bitset.New(100)
	b.Copy(a)
	want := a.Size()
	got := b.Size()
	if got != want {
		t.Errorf("copy: size mis-match got %d want %d", got, want)
	}
}

func TestStringer(t *testing.T) {
	in := []uint64{0, 63, 64, 127}
	want := "1000000000000000000000000000000000000000000000000000000000000001" +
		"1000000000000000000000000000000000000000000000000000000000000001"
	b := bitset.New(100)
	for _, v := range in {
		b.Set(v)
	}
	got := b.String()
	if got != want {
		t.Errorf("string: bit pattern mis-match got %s want %s", got, want)
	}
}


bitset.go
package bitset

import "strings"

const bitsPerWord uint64 = 6
const numOfBits uint64 = 64

// Bitset represents a fixed size sequence of N bits.
type Bitset struct {
	set       []uint64
	size      uint64
	trueCount uint64
}

// New given the desired number of bits, returns a new instance of a bitset.
func New(bits uint64) *Bitset {
	n := (bits >> bitsPerWord) + 1
	return &Bitset{set: make([]uint64, n), size: n * numOfBits}
}

// wordIndex given a position calculates the index in the bitset.
func (b *Bitset) wordIndex(p uint64) uint64 {
	return (p >> bitsPerWord)
}

// posIndex given a position and a word index, locates the target bit in that
// word index.
func (b *Bitset) posIndex(p, n uint64) uint64 {
	return (p - (n * numOfBits))
}

// checkBounds determines if 'n' is greater than the number of bits in the
// bit set. if it is, the method panics.
func (b *Bitset) checkBounds(n uint64) {
	if n > b.size {
		panic("index out of bounds")
	}
}

// Size returns the number of bits that the bitset can hold. The size is a
// a multiple of 64.
func (b *Bitset) Size() uint64 {
	return b.size
}

// Set sets the bit at a given position to 1 (true), the method panics if the
// postion to set, is greater than the number of bits in the bitset.
func (b *Bitset) Set(p uint64) {
	b.checkBounds(p)
	n := b.wordIndex(p)
	b.set[n] |= (1 << b.posIndex(p, n))
	b.trueCount++
}

// Reset sets the bit at a given position to 0 (false), the method panics if
// the postion to reset, is greater than the number of bits in the bitset.
func (b *Bitset) Reset(p uint64) {
	b.checkBounds(p)
	n := b.wordIndex(p)
	if b.Test(p) {
		b.set[n] &= ^(1 << b.posIndex(p, n))
		b.trueCount--
	}
}

// Test returns the value of a bit at a given position. true if the bit is set
// to 1, or false if the bit is set to 0. The method panics if the postion to
// test, is greater than the number of bits in the bitset.
func (b *Bitset) Test(p uint64) bool {
	b.checkBounds(p)
	n := b.wordIndex(p)
	t := b.set[n] & (1 << b.posIndex(p, n))
	if t == 0 {
		return false
	}
	return true
}

// Flip inverts the value of a bit at a given position. the method panics if
// the postion to flip, is greater than the number of bits in the bitset.
func (b *Bitset) Flip(p uint64) {
	b.checkBounds(p)
	n := b.wordIndex(p)
	b.set[n] ^= (1 << b.posIndex(p, n))
	if b.Test(p) {
		b.trueCount++
	} else {
		b.trueCount--
	}
}

// All tests if all the bits in the bitset are set to true.
func (b *Bitset) All() bool {
	var i uint64
	for ; i < b.size; i++ {
		if !b.Test(i) {
			return false
		}
	}
	return true
}

// Any tests if any of the bits in the bitset are set to true.
func (b *Bitset) Any() bool {
	var i uint64
	for ; i < b.size; i++ {
		if b.Test(i) {
			return true
		}
	}
	return false
}

// None tests if none of the bits in the bitset are set to true.
func (b *Bitset) None() bool {
	var i uint64
	for ; i < b.size; i++ {
		if b.Test(i) {
			return false
		}
	}
	return true
}

// TrueCount returns the number of bits in the bitset set to 1 (true).
func (b *Bitset) TrueCount() uint64 {
	return b.trueCount
}

// Copy makes this bitset an exact independant copy of the argument bitset.
func (b *Bitset) Copy(a *Bitset) {
	b.set = nil
	b.set = make([]uint64, len(a.set))
	b.size = a.size
	b.trueCount = a.trueCount
	copy(b.set, a.set)
}

// String returns a string representation of the bitset. The length of the
// string will be equal to bitset.Size(). Matches the Stringer interface.
func (b *Bitset) String() string {
	n := ((b.size - 1) >> bitsPerWord) + 1
	s := make([]string, n*numOfBits)
	var i uint64
	var j uint64
	for ; i < n; i++ {
		x := b.set[i]
		for j = 0; j < 64; j++ {
			if (x & (1 << j)) > 0 {
				s = append(s, "1")
			} else {
				s = append(s, "0")
			}
		}
	}
	return strings.Join(s, "")
}


doc.go
/*
Package bitset represents a fixed size sequence of N bits.

Lets create a new bitset:

    package main

    import (
        "fmt"
        "github.com/umahmood/bitset"
    )

    func main() {
        bits := bitset.New(64)
        ...
    }

This will allocate 64 bits all initialized to zero:

    0000000000000000000000000000000000000000000000000000000000000000

Note: The desired size is rounded upto be a muliple of 64. i.e:

    bitset.New(50)  // Size is rounded upto 64.
    bitset.New(100) // Size is rounded upto 128.
    bitset.New(150) // Size is rounded upto 192.

To set a bit:

    bits.Set(42)

Note: The Most Significant Bit (MSB) is at index zero, so performing:

    bits.Set(0)

Will set the bit at index zero and will result in the bit pattern:

    1000000000000000000000000000000000000000000000000000000000000000
    ↑ msb

Reset a bit:

    bits.Reset(42)

Flip a bit:

    bits.Flip(28)

Test a bit:

    if bits.Test(10) {
        fmt.Println("bit a position 10 set to 1 (true).")
    } else {
        fmt.Println("bit a position 10 set to 0 (false).")
    }

Size of the bitset:

    n := bits.Size()

Check if all the bits in the bitset are set to 1.

    if bits.All() {...}

Check if any of the bits in the bitset are set to 1.

    if bits.Any() {...}

Check if none of the bits in the bitset are set to 1.

    if bits.None() {...}

Get the number of bits in the bitset set to 1.

    n := bits.TrueCount()

Make a copy of a bitset:

    a := bitset.New(200)
    a.Set(42)
    b := bitset.New(100)
    b.Copy(a)
    b.Reset(42)
    a.Test(42) // true
    b.Test(42) // false

Printing a bitset:

    fmt.Println(bits.String())
    // or
    fmt.Println(bits)

*/
package bitset
